src/tree/index.tsx

```diff
@@ -0,0 +1,415 @@
+
+export interface TreeNodeOption {
+	id: string;
+	parent: string;
+	value: string;
+	hasChildren: boolean;
+}
+
+export interface TreeProperties {
+	checkable?: boolean;
+	selectable?: boolean;
+	checked?: string[];
+	expanded?: string[];
+	initialChecked?: string[];
+	initialExpanded?: string[];
+	selectedNode?: string;
+	disabledNodes?: string[];
+	parentSelection?: boolean;
+	onValue?(id: string): void;
+	onCheck?(id: string[]): void;
+	onExpand?(id: string[]): void;
+}
+
+export interface TreeChildren {
+	(node: TreeNodeOption): RenderResult;
+}
+
+interface TreeCache {
+	activeNode?: string;
+	selectedNode?: string;
+	controlledExpandedNodes: string[];
+	controlledCheckedNodes: string[];
+	checked: string[];
+	expanded: string[];
+}
+
+const resource = createResourceMiddleware<TreeNodeOption>();
+const icache = createICacheMiddleware<TreeCache>();
+const factory = create({ theme, icache, diffProperty, focus, resource })
+	.properties<TreeProperties>()
+	.children<TreeChildren | undefined>();
+
+export default factory(function({
+	id,
+	middleware: { theme, icache, diffProperty, focus, resource },
+	properties,
+	children
+}) {
+	diffProperty(
+		'selectedNode',
+		properties,
+		({ selectedNode: current }, { selectedNode: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('selectedNode', next);
+			}
+		}
+	);
+	diffProperty('expanded', properties, ({ expanded: current }, { expanded: next }) => {
+		if ((current || next) && current !== next) {
+			icache.set('controlledExpandedNodes', next || []);
+		}
+	});
+	diffProperty('checked', properties, ({ checked: current }, { checked: next }) => {
+		if ((current || next) && current !== next) {
+			icache.set('controlledCheckedNodes', next || []);
+		}
+	});
+	diffProperty(
+		'initialChecked',
+		properties,
+		({ initialChecked: current }, { initialChecked: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('checked', next || []);
+			}
+		}
+	);
+	diffProperty(
+		'initialExpanded',
+		properties,
+		({ initialExpanded: current }, { initialExpanded: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('expanded', next || []);
+			}
+		}
+	);
+
+	const { getOrRead, createOptions, isLoading, meta } = resource;
+	const {
+		checkable = false,
+		selectable = false,
+		onValue,
+		onCheck,
+		onExpand,
+		disabledNodes,
+		resource: { template },
+		parentSelection = false
+	} = properties();
+	const classes = theme.classes(css);
+	const defaultRenderer = (n: TreeNodeOption) => n.value;
+	const [itemRenderer] = children();
+
+	const activeNode = icache.get('activeNode');
+	const selectedNode = icache.get('selectedNode');
+	const controlledExpandedNodes = icache.get('controlledExpandedNodes');
+	const controlledCheckedNodes = icache.get('controlledCheckedNodes');
+	const expandedNodes = icache.getOrSet('expanded', []);
+	const checkedNodes = icache.getOrSet('checked', []);
```

Comment:
you should be able to determine if you're using the `icache` or properties to the controlled values and simply set them as `consts` here so that you don't need the `if controlledCheckedNodes` etc statements in the rest of the render function.

Reply:
something like the below, taken from `text-input`.

``` tsx
let { value } = properties();
const { initialValue } = properties();

if (value === undefined) {
		value = icache.get('value');
		const existingInitialValue = icache.get('initialValue');

		if (initialValue !== existingInitialValue) {
			icache.set('value', initialValue);
			icache.set('initialValue', initialValue);
			value = initialValue;
		}
	}
```